presto-main/src/main/java/com/facebook/presto/connector/informationSchema/InformationSchemaMetadata.java

```diff
@@ -189,11 +203,52 @@ public ColumnMetadata getColumnMetadata(ConnectorSession session, ConnectorTable
     @Override
     public List<ConnectorTableLayoutResult> getTableLayouts(ConnectorSession session, ConnectorTableHandle table, Constraint<ColumnHandle> constraint, Optional<Set<ColumnHandle>> desiredColumns)
     {
-        InformationSchemaTableHandle handle = (InformationSchemaTableHandle) table;
-        ConnectorTableLayout layout = new ConnectorTableLayout(new InformationSchemaTableLayoutHandle(handle, constraint.getSummary()));
+        InformationSchemaTableHandle handle = checkTableHandle(table);
+
+        TupleDomain<ColumnHandle> predicate = constraint.getSummary();
+        Optional<Stream<Map<ColumnHandle, NullableValue>>> partitions = getPartitions(session, handle.getSchemaTableName(), handle.getCatalogName(), predicate);
+        if (partitions.isPresent()) {
+            List<TupleDomain<ColumnHandle>> partitionsTupleDomains = partitions.get()
+                    .filter(constraint.predicate())
+                    .map(TupleDomain::fromFixedValues)
+                    .collect(toImmutableList());
+            if (partitionsTupleDomains.isEmpty()) {
+                return ImmutableList.of();
+            }
+            predicate = predicate.intersect(TupleDomain.columnWiseUnion(partitionsTupleDomains));
+        }
+
+        ConnectorTableLayout layout = new ConnectorTableLayout(
+                new InformationSchemaTableLayoutHandle(handle, predicate),
+                Optional.empty(),
+                predicate,
+                Optional.empty(),
+                Optional.empty(),
+                Optional.empty(),
+                ImmutableList.of());
         return ImmutableList.of(new ConnectorTableLayoutResult(layout, constraint.getSummary()));
     }
 
+    private Optional<Stream<Map<ColumnHandle, NullableValue>>> getPartitions(ConnectorSession connectorSession, SchemaTableName schemaTableName, String catalogName, TupleDomain<ColumnHandle> constraint)
+    {
+        if (TABLE_COLUMNS.equals(schemaTableName)) {
+            InformationSchemaColumnHandle catalogColumnHandle = new InformationSchemaColumnHandle("table_catalog");
+            InformationSchemaColumnHandle schemaColumnHandle = new InformationSchemaColumnHandle("table_schema");
+            InformationSchemaColumnHandle tableColumnHandle = new InformationSchemaColumnHandle("table_name");
+            Optional<String> schema = stringFilter(constraint, schemaColumnHandle);
```

Comment:
can constraint be `none`?

Reply:
I do not think so, see `com/facebook/presto/metadata/MetadataManager.java:333`
In case of `none` the execution should not reach here.